File: 2024/jl/day02.jl
# Day 2: Red-Nosed Reports

include("util.jl")

using .Util

lines = Util.getinput("day02.txt")

checksorted(d) = (issorted(d) || issorted(d, rev=true))
checkdiff(d) = all(y->abs(y) ∈ 1:3, diff(d))

allok(d) = checksorted(d) && checkdiff(d)
function star1()
	total = 0
	for d ∈ lines
		d = parse.(Int64, split(d))
		if allok(d)
			total += 1
		end
	end

	println("- $total")
end

function star2()
	total = 0
	for d ∈ lines
		d = parse.(Int64, split(d))
		for i ∈ eachindex(d)
			ds = deleteat!(copy(d), i)
			if allok(ds)
				total += 1
				break
			end
		end
	end

	println("- $total")
end

star1()
star2()

File: 2024/py/src/day02.py
# Day 2: Red-Nosed Reports

from src import util


def diff(a):
    return [a[x] - a[x + 1] for x in range(len(a) - 1)]


DATA = util.input_lines("day02.txt")


def star1():
    safe = 0
    for line in DATA:
        if not line:
            break

        dline = diff([int(x) for x in line.split()])
        if (all(x < 0 for x in dline) or all(x > 0 for x in dline)) and abs(
            max(dline, key=abs)
        ) <= 3:
            safe += 1

    return safe


def star2():
    safe = 0
    for line in DATA:
        if not line:
            break

        line = [int(x) for x in line.split()]
        for i in range(len(line)):
            nl = line[:i] + line[i + 1 :]
            dline = diff(nl)
            if (all(x < 0 for x in dline) or all(x > 0 for x in dline)) and abs(
                max(dline, key=abs)
            ) <= 3:
                safe += 1
                break

    return safe


util.run_day(2, star1, star2)
